Wangscape/noise/module/ReseedableOps.cpp

```diff
@@ -0,0 +1,502 @@
+#include "ReseedableOps.h"
+#include "Exp.h"
+#include "Pow.h"
+#include "Gradient.h"
+#include "NormLPQ.h"
+#include "CornerCombinerBase.h"
+
+namespace noise
+{
+namespace module
+{
+
+ReseedablePtr operator+(ReseedablePtr left, ReseedablePtr right)
+{
+    auto add_p = std::make_shared<Reseedable<Add>>();
+    add_p->getModule().SetSourceModule(0, left->getModule());
+    add_p->getModule().SetSourceModule(1, right->getModule());
+
+    return makeModuleGroup({
+        {"source0", left},
+        {"source1", right},
+        {DEFAULT_MODULE_GROUP_OUT, add_p}
+    });
+}
+
+ReseedablePtr operator+(ReseedablePtr left, double right)
+{
+    auto scale_bias_p = std::make_shared<Reseedable<ScaleBias>>();
+    scale_bias_p->module.SetSourceModule(0, left->getModule());
+    scale_bias_p->module.SetScale(1.);
+    scale_bias_p->module.SetBias(right);
+
+    return makeModuleGroup({
+        {"source", left},
+        {DEFAULT_MODULE_GROUP_OUT, scale_bias_p}
+    });
+}
+
+ReseedablePtr operator+(double left, ReseedablePtr right)
+{
+    return right + left;
+}
+
+ReseedablePtr& operator+=(ReseedablePtr& left, ReseedablePtr right)
+{
+    return (left = left + right);
+}
+
+ReseedablePtr& operator+=(ReseedablePtr& left, double right)
+{
+    return (left = left + right);
+}
+
+ReseedablePtr operator-(ReseedablePtr left, ReseedablePtr right)
+{
+    return left + -right;
+}
+
+ReseedablePtr operator-(ReseedablePtr left, double right)
+{
+    return left + -right;
+}
+
+ReseedablePtr operator-(double left, ReseedablePtr right)
+{
+    return left + -right;
+}
+
+ReseedablePtr operator-(ReseedablePtr source)
+{
+    return invert(source);
+}
+
+ReseedablePtr invert(ReseedablePtr source)
+{
+    auto invert_p = std::make_shared<Reseedable<Invert>>();
+    invert_p->module.SetSourceModule(0, source->getModule());
+
+    return makeModuleGroup({
+        {"source", source},
+        {DEFAULT_MODULE_GROUP_OUT, invert_p}
+    });
+}
+
+ReseedablePtr& operator-=(ReseedablePtr& left, ReseedablePtr right)
+{
+    return (left = left + -right);
+}
+
+ReseedablePtr& operator-=(ReseedablePtr& left, double right)
```

Comment:
In the old scheme for module networks, with `ModuleGroup`s containing `ReseedablePtr`s, there might have been some semantic differences between passing a `shared_ptr` and a `shared_ptr&`. But all of these operators will be deleted, so it's not worth worrying about.